net-p2p/bitcoin-core/bitcoin-core-25.0.ebuild

```diff
@@ -0,0 +1,260 @@
+
+	sed -e 's/^\(complete -F _bitcoind\b\).*$/\1'"$(usev daemon ' bitcoind')$(usev qt5 ' bitcoin-qt')/" \
+		-i contrib/completions/bash/bitcoind.bash-completion || die
+}
+
+src_configure() {
+	local wallet ; if use berkdb || use sqlite ; then wallet=enable ; else wallet=disable ; fi
+	local myeconfargs=(
+		--disable-static
+		--${wallet}-wallet
+		$(use_with sqlite)
+		$(use_with berkdb bdb)
+		$(use_enable systemtap usdt)
+		$(use_with upnp miniupnpc)
+		$(use_with nat-pmp natpmp)
+		$(use_enable test tests)
+		--disable-bench
+		--disable-fuzz{,-binary}
+		$(use_with qrcode qrencode)
+		--disable-ccache
+		$(use_enable asm)
+		$(use_enable zeromq zmq)
+		$(use_enable man)
+		$(use_enable external-signer)
+		--with-utils
+		$(use_enable {bitcoin,util}-cli)
+		--enable-util-tx
+		--${wallet}-util-wallet
+		--disable-util-util
+		$(use_with libs)
+		$(use_with daemon)
+		$(use_with qt5 gui qt5)
+		$(use_with dbus qtdbus)
+		$(use_with system-leveldb)
+		$(use_with system-libsecp256k1)
+	)
+	econf "${myeconfargs[@]}"
+}
+
+src_install() {
+	use external-signer && DOCS+=( doc/external-signer.md )
+	use berkdb || use sqlite && DOCS+=( doc/managing-wallets.md )
+	use libs && DOCS+=( doc/shared-libraries.md )
+	use systemtap && DOCS+=( doc/tracing.md )
+	use zeromq && DOCS+=( doc/zmq.md )
+
+	default
+
+	! use libs || find "${ED}" -name '*.la' -delete || die
+	! use test || rm -f -- "${ED}"/usr/bin/test_bitcoin{,-qt} || die
+
+	newbashcomp contrib/completions/bash/bitcoin-tx.bash-completion bitcoin-tx
+	use bitcoin-cli && newbashcomp contrib/completions/bash/bitcoin-cli.bash-completion bitcoin-cli
+	use daemon || use qt5 && newbashcomp contrib/completions/bash/bitcoind.bash-completion bitcoind
+
+	if use daemon ; then
+		insinto /etc/bitcoin
+		sed -ne 's/ To use, copy this file$//p;Tp;:0;n;/save the file\.$/!b0;d;:p;p' \
+			share/examples/bitcoin.conf >"${ED}/etc/bitcoin/bitcoin.conf" || die
+		fowners bitcoin:bitcoin /etc/bitcoin/bitcoin.conf
+		fperms 0660 /etc/bitcoin/bitcoin.conf
+
+		newconfd contrib/init/bitcoind.openrcconf bitcoind
+		newinitd "${FILESDIR}/bitcoind.openrc" bitcoind
+		systemd_newunit contrib/init/bitcoind.service bitcoind.service
+
+		keepdir /var/lib/bitcoind
+		fperms 0750 /var/lib/bitcoind
+		fowners bitcoin:bitcoin /var/lib/bitcoind
+		dosym -r {/etc/bitcoin,/var/lib/bitcoind}/bitcoin.conf
+
+		insinto /etc/logrotate.d
+		newins "${FILESDIR}/bitcoind.logrotate-r1" bitcoind
+	fi
+
+	if use qt5 ; then
+		insinto /usr/share/icons/hicolor/scalable/apps
+		newins src/qt/res/src/bitcoin.svg bitcoin128.svg
+
+		domenu "${FILESDIR}/org.bitcoin.bitcoin-qt.desktop"
+
+		if use kde ; then
+			insinto /usr/share/kservices5
+			doins "${FILESDIR}/bitcoin-qt.protocol"
+		fi
+	fi
+
+	if use examples ; then
+		docinto examples
+		dodoc -r contrib/{linearize,qos}
+		use zeromq && dodoc -r contrib/zmq
+	fi
+}
+
+pkg_preinst() {
+	if use daemon && [[ -d "${EROOT}/var/lib/bitcoin/.bitcoin" ]] ; then
```

Comment:
<sup>Continuing our discussion from https://github.com/gentoo/gentoo/pull/32978#discussion_r1335877482 here</sup>

> Can you recall any gotchas that arose in that migration?

One was and is https://bugs.gentoo.org/856691#c11. But I assume the acct-user home change does not apply here.

After sleeping about this, I personally would probably go a slightly more defensive route. Especially, since this is bitcoin, where data migration potentially means bitcoin wallet migration, which is a very sensitive topic with regard to data loss.

So I would
1. Check that the destination has enough free space
2. Check that there is no running bitcoin instance
3. Copy the data, instead of moving it
4. Factory out the migration in an extra script installed by the ebuild and have pkg_preinst only inform the user that the directory should be migrated

But let's give others some time to express their opinion.

Reply:
And note that we do not need to tackle the datadir migration together with the bump. Bumping first, and migrate (/add a migration script) later in a revision bump would also be an option.

Reply:
> But I assume the acct-user home change does not apply here.

Correct. `acct-user/bitcoin` does not specify a home directory, so it gets `/dev/null`.

> * Check that the destination has enough free space

Not an issue if the move is really a move, not a copy+remove.

> * Check that there is no running bitcoin instance

This _could_ be a legitimate concern, as there will be a brief window of time (between executing `pkg_preinst()` and merging the image to the live filesystem) in which `/var/lib/bitcoin/.bitcoin` does not exist, either as a directory or as a symlink to a directory. If bitcoind attempts to open any files beneath that path during that brief window, it would fail. I do not know whether Bitcoin Core is smart/defensive enough to hold onto a file descriptor to its data directory and pass it to `openat(3p)` to avoid issues of its data directory's being moved while it's running, but given the cross-platform nature of the project, my bet would be it isn't, so any attempts to open new blockchain data or undo files or new LevelDB logs or shards during the transition window would fail. Not good.

> * Copy the data, instead of moving it (but `mv --no-copy`, as you currently do, could be an option too)

NACK. Copying is not acceptable. The blockchain is over half a terabyte in size.

> * Factor out the migration in an extra script installed by the ebuild and have pkg_preinst only inform the user that the directory should be migrated

This is probably the direction I would go, but I dislike installing a script. Any reason not to use a `pkg_configure()` in the ebuild to perform the migration (if applicable, or else it becomes a no-op)? The `pkg_postinst()` would inform the user of the pending migration and the need to run `emerge --config ${CATEGORY}/${PN}` after shutting down any running instances of bitcoind that are using `/var/lib/bitcoin/.bitcoin` as their data directory. The `pkg_configure()` would also perform defensive checks that the directory is not in use, though those might not be 100% fool-proof, thus the need for manual action by the user.

Reply:
> And note that we do not need to tackle the datadir migration together with the bump.

True, although I loathe when large packages do trivial revbumps that don't affect the installed binaries in any way, as compilation time is not negligible. I would rather get it all done in one shot, even if the data-dir migration has to be a manual step for the user.

Reply:
Another alternative is to perform the migration in a `start_pre()` function of the `/etc/init.d/bitcoind` runscript. Then it's still automatic, but it's deferred until we know that the bitcoind service is not running. However, this only handles the OpenRC case, and I do not care to educate myself on hacking the abomination that is SystemD. Further, it would miss the case that someone is using `/var/lib/bitcoin/.bitcoin` as the data dir of a bitcoind instance that is not under management of the system service manager. So probably this alternative is a no-go.

Reply:
Okay, new logic in da8a1e902f:

* In `pkg_preinst()`:
    * If `${EROOT}/var/lib/bitcoin/.bitcoin` is a symlink, then we retain the symlink for compatibility. (The symlink should only disappear when the user intentionally removes it or when they uninstall the package.)
    * If `${EROOT}/var/lib/bitcoin/.bitcoin` is a directory (and is not a symlink), and if `${EROOT}/var/lib/bitcoind` does not exist or is a symlink, then we:
        * emit a warning about the pending migration;
        * install our keepfile into the deprecated location instead; and 
        * install a symlink at the new location, pointing at the deprecated location, so that the new service runscript/unit will work.
* In `pkg_config()`:
    * If `${EROOT}/var/lib/bitcoin/.bitcoin` is a directory (and is not a symlink), and if `${EROOT}/var/lib/bitcoind` does not exist or is a symlink, then we:
        * collect up all the targets of relative symlinks beneath the data dir that point outside the data dir;
        * remove any symlink that might be at the new location;
        * move (without copying) the data dir to the new location;
        * rewrite the symlinks that we collected earlier so that their relative targets are still correct, now that the symlinks are one level higher in the directory hierarchy;
        * create a symlink at the deprecated location, pointing at the new location, in case the user has any other configs or scripts still referencing the old location.

Reply:
I think checking if a migration is necessary in `pkg_preinst` and asking the user to explicitly trigger the migration via `emerge --config` is the right approach. Thanks.